### ISSUE TITLE
How to get the attention score while forcing the output of the decoder?

### ISSUE THREAD

**Issue by anoidgit** [OpenNMT/OpenNMT#130]
I need to evaluate the alignment of models, I have traced the code here [`function Decoder:forwardAndApply(batch, encoderStates, context, func)`](https://github.com/OpenNMT/OpenNMT/blob/master/onmt/modules/Decoder.lua#L261) and I know how to force the output of the decoder, but I'm not clear about how to get the output of attention after the decoder forwarded [`local outputs = self:net(t):forward(inputs)`](https://github.com/OpenNMT/OpenNMT/blob/master/onmt/modules/Decoder.lua#L237) in [`function Decoder:forwardOne(input, prevStates, context, prevOut, t)`](https://github.com/OpenNMT/OpenNMT/blob/master/onmt/modules/Decoder.lua#L209)
Thank you if you could help me.

**Comment by anoidgit** [comment]
I do not quite understand why, but seems maybe I could get the score from `self.decoder.softmaxAttn.output`

**Comment by guillaumekln** [comment]
Yes you can get the scores here. Be careful though, it contains the attention vector for each beam of each batch.

Was retrieving attention vectors at a higher level (#128) not sufficient?

**Comment by anoidgit** [comment]
Thank you for your help, I want to get the attention score will force the decoder to decode the gold translation, and I'm not sure if I could get the attention score of the gold translation in that higher level. @guillaumekln

**Comment by guillaumekln** [comment]
Getting attention vectors from the gold data is not made easy currently. Ideally, they should be retrieved when [computing the gold scores](https://github.com/OpenNMT/OpenNMT/blob/master/onmt/translate/Translator.lua#L174). Here is a suggestion:

You rightly pinpointed the decoder forward:

```lua
local outputs = self:net(t):forward(inputs)
```

To dump the attention of the gold data, you could add directly after:

```lua
if inputSize == 1 then
  local file = io.open('gold-attn.txt', 'a')
  if t == 1 then
    file:write('\n')
  end
  for i = 1, self.softmaxAttn.output:size(2) do
    if i > 1 then
      file:write(' ')
    end
    file:write(self.softmaxAttn.output[1][i])
  end
  file:write('\n')
end
```

which will append the current attention vector to the `gold-attn.txt` file and add an empty line before every new sentences. You can then invoke `translate.lua` with your gold data and `-batch_size 1` to construct the file.

**Comment by anoidgit** [comment]
Thank you for your help, I have been confusion with this problem for a day. I hope some time I could help you when you need. @guillaumekln

**Comment by shahbazsyed** [comment]
@anoidgit Can you kindly tell me how did you force the output of the decoder? By forcing, did you mean you made the decoder output a set of words from the source in the final translation?

**Comment by anoidgit** [comment]
Hi, If you already have correct translations and you want to analyze alignment quality or something else, you could run `translate.lua` with option `-tgt your_correct_translation_file`, then `translate.lua` will emulate what will happened to give the correct translation result in your `your_correct_translation_file`. Sorry for my poor english. @shahbazsyed

**Comment by shahbazsyed** [comment]
@anoidgit Thanks for the suggestion. I already do this when I translate my test set. Is it possible to get the words with highest attention from the source during this phase? I looked into the code a bit but could not figure out where exactly is the place where source attention is being calculated.

Another thing I would like to know is if I have a set of source words that I need to be present in the translation by the decoder for a given source->target pair, how can I achieve this either in the training or the translation step? 

Thanks !

**Comment by anoidgit** [comment]
I do not know how to present a set of source words in translation, maybe coverage will help in some cases. You could get the words with highest attention score with `self.decoder.softmaxAttn.output` and your target vocabulary. @shahbazsyed

**Comment by anoidgit** [comment]
`-replace_unk` option could get the source words with highest attention score for "<unk>", maybe you could check the code related with this option. @shahbazsyed
